engine-standalone-storage/src/engine_state.rs

```diff
@@ -0,0 +1,136 @@
+        }
+    }
+}
+
+impl<'a> StorageIntermediate for EngineStorageValue<'a> {
+    fn len(&self) -> usize {
+        self.as_ref().len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.as_ref().is_empty()
+    }
+
+    fn copy_to_slice(&self, buffer: &mut [u8]) {
+        buffer.copy_from_slice(self.as_ref())
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct Storage<'db, 'input, 'output> {
+    input: &'input [u8],
+    bound_block_height: u64,
+    bound_tx_position: u16,
+    transaction_diff: &'output RefCell<Diff>,
+    output: &'output Cell<Vec<u8>>,
+    db: &'db DB,
+}
+
+impl<'db, 'input, 'output> Storage<'db, 'input, 'output> {
+    pub fn new(
+        input: &'input [u8],
+        bound_block_height: u64,
+        bound_tx_position: u16,
+        transaction_diff: &'output RefCell<Diff>,
+        output: &'output Cell<Vec<u8>>,
+        db: &'db DB,
+    ) -> Self {
+        Self {
+            input,
+            bound_block_height,
+            bound_tx_position,
+            transaction_diff,
+            output,
+            db,
+        }
+    }
+
+    pub fn get_transaction_diff(&self) -> Diff {
+        self.transaction_diff.borrow().clone()
+    }
+
+    fn construct_engine_read(&self, key: &[u8]) -> rocksdb::ReadOptions {
+        let upper_bound =
+            super::construct_engine_key(key, self.bound_block_height, self.bound_tx_position);
+        let lower_bound = super::construct_storage_key(StoragePrefix::Engine, key);
+        let mut opt = rocksdb::ReadOptions::default();
+        opt.set_iterate_upper_bound(upper_bound);
+        opt.set_iterate_lower_bound(lower_bound);
+        opt
+    }
+}
+
+impl<'db, 'input: 'db, 'output: 'db> IO for Storage<'db, 'input, 'output> {
+    type StorageValue = EngineStorageValue<'db>;
+
+    fn read_input(&self) -> Self::StorageValue {
+        EngineStorageValue::Slice(self.input)
+    }
+
+    fn return_output(&mut self, value: &[u8]) {
+        self.output.set(value.to_vec())
+    }
+
+    fn read_storage(&self, key: &[u8]) -> Option<Self::StorageValue> {
+        if let Some(diff) = self.transaction_diff.borrow().get(key) {
+            return diff
+                .value()
+                .map(|bytes| EngineStorageValue::Vec(bytes.to_vec()));
+        }
+
+        let opt = self.construct_engine_read(key);
+        let mut iter = self.db.iterator_opt(rocksdb::IteratorMode::End, opt);
+        let value = iter
+            .next()
+            .map(|(_, value)| DiffValue::try_from_bytes(&value).unwrap())?;
+        value.take_value().map(EngineStorageValue::Vec)
+    }
+
+    fn storage_has_key(&self, key: &[u8]) -> bool {
+        self.read_storage(key).is_some()
+    }
+
+    fn write_storage(&mut self, key: &[u8], value: &[u8]) -> Option<Self::StorageValue> {
+        let original_value = self.read_storage(key);
+
+        self.transaction_diff
+            .borrow_mut()
+            .modify(key.to_vec(), value.to_vec());
+
+        original_value
+    }
```

Comment:
Should we allow some option to write to storage without reading evicted value! I think this feature is not always necessary, and reading from storage can be slow if the value is not in the diff right?

Reply:
Same with `remove_storage`.

Reply:
I see this is part of IO trait, and probably that involve a bigger change. 

Don't take this comment as blocking, just something that passed through my mind while reading the code.

Reply:
The interface for the `IO` trait is based on the [one from `NEAR` itself](https://nomicon.io/RuntimeSpec/Components/BindingsSpec/TrieAPI.html) where `write` also takes a register to store the old value.

I agree the current implementation is more overhead when the old value is not used. An alternative implementation is the following:

We add another variant to `EngineStorageValue` called `LazyDBRead` which holds the key. And its `AsRef` implementation actually performs the read to return the value. Then for `write_storage` and `remove_storage` we check if the key is in the diff, if so we capture the value as we do now, but if not this `LazyDBRead` is returned insteasd. This means no DB read is done unless the returned value is used. It works because we do not modify the DB during the lifetime of this `IO` instance (writes only change the Diff), so we can technically perform the write before reading the old value from the DB.

What do you think about that? It is more efficient (writes only require allocating memory to store the key, not DB reads unless the value is used), but more complex also. I'm not sure what level of optimization we need for the standalone engine. It probably has an inherent advantage in terms of performance because it does not require wasm interpretation.

Reply:
This is an excellent suggestion, though not strictly necessary for now. I'm ok skipping it.